File: LeetCode/222完全二叉树的节点个数.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
void calc(struct TreeNode *root, int *cnt) {
    if (root == NULL) return ;
    calc(root->left, cnt);
    *cnt += 1;
    calc(root->right, cnt);
    return ;
}

int countNodes(struct TreeNode* root){
    int cnt = 0;
    calc(root, &cnt);
    return cnt;
}

File: LeetCode/222完全二叉树的节点个数.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // public int countNodes(TreeNode root) {
    //     if (root == null) return 0;
    //     return 1 + countNodes(root.left) + countNodes(root.right);
    // }

    public int countNodes(TreeNode root) {
        if(root == null) {
            return 0;
        }
        int leftDepth = getDepth(root.left);
        int rightDepth = getDepth(root.right);
        if (leftDepth == rightDepth) {
            return (1 << leftDepth) + countNodes(root.right);
        } else {
            return (1 << rightDepth) + countNodes(root.left);
        }
    }

    private int getDepth(TreeNode root) {
        int depth = 0;
        while (root != null) {
            root = root.left;
            depth++;
        }
        return depth;
    }
}